### PR TITLE
Global counters

### PR DESCRIPTION
**Pull request by dcorbacho**
Introduces global counters using the new [seshat](https://github.com/rabbitmq/seshat) library.
Also adds some missing metrics to the stream plugin.

Requires https://github.com/rabbitmq/osiris/pull/30 & https://github.com/rabbitmq/ra/pull/221

## Types of Changes
- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist
- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

### CODE REVIEW

File: deps/rabbit/src/rabbit_messages_counters.erl [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+init() ->
+    seshat_counters:new_group(queue),
+    seshat_counters:new_group(global),
+    new(global, global, []),
+    ok.
+
+new(Group, Object, Fields) ->
+    %% Some object could have extra metrics, i.e. queue types might have their own counters
+    seshat_counters:new(Group, Object, ?COUNTERS ++ Fields).
```
Comment (kjnilsson) [inline deps/rabbit/src/rabbit_messages_counters.erl:96]: nice I like the meta data stuff for the counter fields

**Review by kjnilsson — CHANGES_REQUESTED** [review]

File: deps/rabbit/src/rabbit_messages_counters.erl [hunk 2]
```diff
@@ -0,0 +1,202 @@
+%% This Source Code Form is subject to the terms of the Mozilla Public
+%% License, v. 2.0. If a copy of the MPL was not distributed with this
+%% file, You can obtain one at https://mozilla.org/MPL/2.0/.
+%%
+%% Copyright (c) 2007-2021 VMware, Inc. or its affiliates.  All rights reserved.
+%%
+
+-module(rabbit_messages_counters).
```
[the remark this answers is not in the document]
Comment (dcorbacho) [reply deps/rabbit/src/rabbit_messages_counters.erl:8]: @gerhard @mkuratczyk ?

[the remark this answers is not in the document]
Comment (gerhard) [reply deps/rabbit/src/rabbit_messages_counters.erl:8]: @kjnilsson + @gerhard:

We should start with more generic counters, and then refine them on a per-protocol basis. We still need the globals to work correctly and independently of the protocol-specific ones.

**Comment by gerhard** [comment]
Superseded by https://github.com/rabbitmq/rabbitmq-server/pull/3127

**Pull request closed without merging**